docs/contributing-architecture.md

```diff
@@ -0,0 +1,476 @@
+stack and generate a stack trace for wasm code. This does have other benefits as
+well such as improving generic sampling profilers when used with Wasmtime.
+
+## Linear Memory
+
+Linear memory in Wasmtime is implemented effectively with `mmap` (or the
+platform equivalent thereof), but there are some subtle nuances that are worth
+pointing out here too. The implementation of linear memory is relatively
+configurable which gives rise to a number of situations that both the runtime
+and generated code need to handle.
+
+First there are a number of properties about linear memory which can be
+configured:
+
+* `wasmtime::Config::static_memory_maximum_size`
+* `wasmtime::Config::static_memory_guard_size`
+* `wasmtime::Config::dynamic_memory_guard_size`
+* `wasmtime::Config::guard_before_linear_memory`
+
+The methods on `Config` have a good bit of documentation to go over some
+nitty-gritty, but the general gist is that Wasmtime has two modes of memory:
+static and dynamic. Static memories represent an address space reservation that
+never moves and pages are committed to represent memory growth. Dynamic
+memories represent allocations where the committed portion exactly matches the
+wasm memory's size and growth happens by allocating a bigger chunk of memory.
+
+The guard size configuration indicates the size of the guard region that
+happens after linear memory. This guard size affects whether generated JIT code
+emits bounds checks or not. Bounds checks are elided if out-of-bounds addresses
+provably encounter the guard pages.
+
+The `guard_before_linear_memory` configuration additionally places guard pages
+in front of linear memory as well as after linear memory (the same size on both
+ends). This is only used to protect against possible Cranelift bugs and
+otherwise serves no purpose.
+
+At the time of this writing Wasmtime only supports WebAssembly with 32-bit
+memories, so a maximum of 4GB in size. Wasmtime has not implemented the
+memory64 proposal from upstream WebAssembly yet.
+
+The defaults for Wasmtime on 64-bit platforms are:
+
+* 4gb static maximum size (meaning all memories are static)
+* 2gb static guard size (meaning all loads/stores with less than 2gb offset
+  don't need bounds checks)
+* Guard pages before linear memory are enabled.
+
+Altogether this means that linear memories result in an 8gb virtual address
+space reservation by default in Wasmtime. With the pooling allocator where we
+know that linear memories are contiguous this results in a 6GB reservation per
+memory because the guard region after one memory is the guard region before the
+next.
+
+## Tables and `externref`
+
+WebAssembly tables contain reference types, currently either `funcref` or
+`externref`. A `funcref` in Wasmtime is represented as `*mut
+VMCallerCheckedAnyfunc` and an `externref` is represented as `*mut
+VMExternRef`. Tables are consequently represented as vectors of pointers.
```

Comment:
An `externref` is actually `VMExternRef`, not `*mut VMExternRef`, and `VMExternRef` is a transparent newtype over `*mut VMExternData`.